### PR TITLE
Modified version of cosmopower_NN

### PR DESCRIPTION
**Pull request by pburger112**
Dear Alessio, 
As you requested, I have closed the first pull request and started a new one. 

1. I updated the cosmopower_NN module
2. Wrote a short jupyter notebook illustrating the new feature.
3. DES 3x2pt models were uploaded as input features and corresponding parameters. 

I saved the training data to the notebook directory. Alternatively, I could send you the data, and you could add a Google Drive link, as in the other notebook file.

Cheers
Pierre

### CODE REVIEW

**Comment by dpiras** [comment]
Dear @pburger112, many thanks for the PR! Echoing what @alessiospuriomancini mentioned in the [closed PR](https://github.com/alessiospuriomancini/cosmopower/pull/26), it sounds like an interesting set of features.

I haven't checked the changes yet, but at a general level it would be good to only have a single method `cosmopower_NN`, as Alessio already mentioned, and as I think you implemented in this PR. One thing that would be nice is not just passing subtraction or division arguments to a new ad-hoc method, but having a function encapsulating the preprocessing, being as general as possible (as long as it's invertible). Do you think you could implement this, andthat this would be useful? In short, what I'm thinking about is that instead of passing a subtraction or division arguments to a new `rescaling` method, we have a single method that simply applies the function `preprocessing` that is passed as input.

Also, I should mention that for negative inputs we found the [PCA+NN](https://github.com/alessiospuriomancini/cosmopower/blob/main/cosmopower/cosmopower_PCAplusNN.py) version of CosmoPower to work quite well - check it out in case you haven't already! 

And thanks again!

**Comment by alessiospuriomancini** [comment]
feel free to update this one @pburger112

**Comment by pburger112** [comment]
Dear both of you, 
I have made the modification Davide asked for. However, my solution is not very elegant. The problem is that the user must provide the preprocessing and postprocessing functions, which undoes the preprocessing for the numpy and TensorFlow methods. 
Furthermore, I wanted to give the option of predefined pre/postprocessing functions (mean, sigma and min, max scaling).
Let me know if you see any improvements.
Cheers
Pierre

File: cosmopower/cosmopower_NN.py
Comment (dpiras) [inline cosmopower/cosmopower_NN.py:213]: Please fix typo

**Comment by dpiras** [comment]
Thank you for the changes @pburger112!

I left one comment about a typo, and I have a couple of more requests. First of all, could you please `git rm` or `gitignore` the `.DS_Store` files, which I think are not needed? Second, I can see why it becomes cumbersome, but at least it provides more flexibility. One important thing in my opinion is that we don't break any backward compatibility, so can you confirm that nothing changes in the current notebooks and scripts, which will run with default parameters and getting the same results?

Finally, I can see you added a new notebook (good) with the corresponding parameter file and pre-trained model. Could you please quickly describe here what is happening in that notebook? I can see it says `3x2pt`, but how is it different from the notebooks we already have?

Thank you so much!

**Comment by dpiras** [comment]
(let me also add that one could use libraries like https://github.com/alvarosg/pynverse to obtain a numerical inverse, but that will always be less reliable than the user providing the postprocessing function, so for now I'd rather stick to the current implementation)

File: cosmopower/cosmopower_NN.py
Comment (pburger112) [reply cosmopower/cosmopower_NN.py:213]: Done!

**Comment by pburger112** [comment]
Hi @dpiras 
I added `.DS_store` to gitignore. Done this the first time, so hopefully, it worked. 

Regarding backward compatibility, yes, I have not touched anything implemented previously and have only added new modules to ensure it can be used as before. 

In the added notebook, I show how the new features can improve the emulation of arbitrary features. To create the `3x2pt` data vectors, I used the DES implementation in cosmosis and varied only Om, Ob, As, and w0. I could also use the power spectra you used in previous notebooks, but the 3x2pt might also be a nice example of the usage of cosmopower. Furthermore, the fact that it works with negative values might be interesting for some users, too. 

Thank you for sharing [https://github.com/alvarosg/pynverse]. I agree that to achieve the best accuracy, it is better if the user also provides the post-processing. Also, I expect that it would make the code much slower, or?

**Comment by dpiras** [comment]
Thank you @pburger112! I took a look and I think a few more edits should be made.

- First of all, you would need to add a * in the `.gitignore` before the `DS_Store`, to ignore all occurrences. You also need to `git rm` the `DS_Store` files to actually remove them.
- The 3x2pt notebook uses some data that is saved in the repository. We would encourage you to actually store these data on e.g. Google Drive and use e.g. `gdown` to download them when needed, to prevent the size of the repository from growing.
- The 3x2pt notebook has a `from cosmopower_NN import cosmopower_NN`, which I think should be changed to `from cosmopower import cosmopower_NN`?
- `As we can see, they vary over a large range of values and have negative values. Therefore, using the log10 of these features is not possible.`: I would add a link to the [PCAplusNN](https://github.com/alessiospuriomancini/cosmopower/blob/main/notebooks/training_notebooks/cosmopower_PCAplusNN_CMB_training.ipynb) training notebook, which has been shown to give good results when data can be negative.
- There are some typos in the comments of the 3x2pt notebook -- coule you please fix them?
- Could you please add some comments before the last two tests, when using the `mean_sigma` and `min_max` methods, explaining what is happening in those cells? Could we also remove the print of `mean` and `sigma` in the former case?
- The y axis on all accuracy plot includes an absolute value, so the code should reflect this. The y limits should be consequently changed.

Once these comments are addressed it should be just a matter of taking a final look. Thanks!

**Comment by pburger112** [comment]
Dear @dpiras 
Thank you for your comments. I should have updated all of them. Please let me know if you need any further typos. 

Regarding the y-axis. I removed the absolute values in the label, as it is interesting to see if the emulator gives biased results. 

All the best
Pierre

**Comment by dpiras** [comment]
Thank you @pburger112! I will take a look at the notebook very soon.

Could you please also `git rm` the two other `DS_Store` files left, namely `notebooks/training_notebooks/.DS_Store` and `notebooks/.DS_Store`?

**Comment by pburger112** [comment]
Thanks for pointing @dpiras . Sorry, I was not seeing them in my ide. But they are removed now.

**Comment by dpiras** [comment]
So sorry for messing so much with this (I should probably do it myself), but your latest commit has re-included the `.DS_Store` file - could you please `git rm` that too? 🙏

**Comment by pburger112** [comment]
No worries, it's my fault. They always show up again :D And annoyingly, I only see them on the website. But they should be hopefully removed now

**Comment by dpiras** [comment]
Thank you so much @pburger112! I took a look at the notebook and modulo the two things I'm listing below it all looks good to me. I'll let @alessiospuriomancini give his final advice!

1. Make sure to either remove or update the reference to the Colab notebook: the current link leads to another notebook on Colab.
2. In themin/max scaling comment, "which we reads" --> "which we read"

**Comment by pburger112** [comment]
Thank you at @dpiras for the comments. I have added those and a little sentence to the readme. Since I never used Colab, I have just removed it.

**Comment by dpiras** [comment]
Thank you @pburger112! The README should have the hyperlink to the notebook, so that people can click on it. Also, would you mind specifying in this README that this notebook is in particular a showcase of the preprocessing features you added, so that it's clear why it's different from the CMB `cosmopowerNN`?

**Comment by pburger112** [comment]
Done! The link does not work yet, as I used a link to the main branch, which is not included in there yet.

**Comment by dpiras** [comment]
Thanks @pburger112! I'll wait for @alessiospuriomancini to give the final approval, looks good to me 👍🏻

**Review by alessiospuriomancini — APPROVED** [review]

**Comment by alessiospuriomancini** [comment]
thank you @pburger112 @dpiras , closing and merging

**Pull request merged**
